File: src/test/java/mnm/mods/tabbychat/test/PatternTests.java
package mnm.mods.tabbychat.test;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.junit.Assert;
import org.junit.Test;

import com.google.common.collect.Lists;

import mnm.mods.tabbychat.util.ChannelPatterns;

public class PatternTests {


    @Test
    public void testAngles() throws Exception {
        test(ChannelPatterns.ANGLES, "<%s> %s");
    }
    @Test
    public void testBraces() throws Exception {
        test(ChannelPatterns.BRACES, "{%s} %s");
    }

    @Test
    public void testBrackets() throws Exception {
        test(ChannelPatterns.BRACKETS, "[%s] %s");
    }
    @Test
    public void testParens() throws Exception {
        test(ChannelPatterns.PARENS, "(%s) %s");
    }

    @Test
    public void testAnglesParens() throws Exception {
        test(ChannelPatterns.ANGLESPARENS, "<(%s) Someone> %s");
    }
    @Test
    public void testAnglesBrackets() throws Exception {
        test(ChannelPatterns.ANGLESBRACKETS, "<[%s] Someone> %s");
    }

    private void test(ChannelPatterns pattern, String format) {
        List<ChanPattern> list = setupChannelPatterns(pattern, format);
        for (ChanPattern chan : list) {
            Assert.assertEquals(chan.channel, chan.testChannel());
        }
    }

    private List<ChanPattern> setupChannelPatterns(ChannelPatterns pattern, String chan) {
        String[] channels = {
                "global",
                "g",
                "local"
        };
        String[] messages = {
                "<JoyJoy> Hello there",
                "[Admin] <MrTCP> ur ba&ned"
        };
        List<ChanPattern> received = Lists.newArrayList();
        for (int i = 0; i < channels.length; i++) {
            for (int j = 0; j < messages.length; j++) {
                String msg = String.format(chan, channels[i], messages[j]);
                received.add(new ChanPattern(pattern, channels[i], msg));
            }
        }
        return received;
    }

    private class ChanPattern {
        final ChannelPatterns pattern;
        final String channel;
        final String message;

        public ChanPattern(ChannelPatterns pattern, String channel, String message) {
            this.pattern = pattern;
            this.channel = channel;
            this.message = message;
        }

        public String testChannel() {
            Pattern pattern = this.pattern.getPattern();
            Matcher matcher = pattern.matcher(message);
            if (matcher.find()) {
                return matcher.group(1);
            }
            return null;
        }

    }
}

File: src/test/kotlin/mnm/mods/tabbychat/test/PatternTests.kt
package mnm.mods.tabbychat.test

import com.google.common.collect.Lists
import mnm.mods.tabbychat.util.ChannelPatterns
import org.junit.Assert
import org.junit.Test

class PatternTests {

    @Test
    fun testAngles() {
        test(ChannelPatterns.ANGLES, "<%s> %s")
    }

    @Test
    @Throws(Exception::class)
    fun testBraces() {
        test(ChannelPatterns.BRACES, "{%s} %s")
    }

    @Test
    fun testBrackets() {
        test(ChannelPatterns.BRACKETS, "[%s] %s")
    }

    @Test
    fun testParens() {
        test(ChannelPatterns.PARENS, "(%s) %s")
    }

    @Test
    fun testAnglesParens() {
        test(ChannelPatterns.ANGLESPARENS, "<(%s) Someone> %s")
    }

    @Test
    fun testAnglesBrackets() {
        test(ChannelPatterns.ANGLESBRACKETS, "<[%s] Someone> %s")
    }

    private fun test(pattern: ChannelPatterns, format: String) {
        val list = setupChannelPatterns(pattern, format)
        for (chan in list) {
            Assert.assertEquals(chan.channel, chan.testChannel())
        }
    }

    private fun setupChannelPatterns(pattern: ChannelPatterns, chan: String): List<ChanPattern> {
        val channels = arrayOf("global", "g", "local")
        val messages = arrayOf("<JoyJoy> Hello there", "[Admin] <MrTCP> ur ba&ned")
        val received = Lists.newArrayList<ChanPattern>()
        for (channel in channels) {
            for (message in messages) {
                val msg = String.format(chan, channel, message)
                received.add(ChanPattern(pattern, channel, msg))
            }
        }
        return received
    }

    private inner class ChanPattern(val pattern: ChannelPatterns, val channel: String, val message: String) {

        fun testChannel(): String? {
            val pattern = this.pattern.pattern
            return pattern.find(message)?.groups?.get(1)?.value
        }

    }
}
